inlong-dataproxy/dataproxy-source/src/main/java/org/apache/inlong/dataproxy/sink/PulsarSink.java

```diff
@@ -283,21 +291,28 @@ public void diffSetPublish(PulsarClientService pulsarClientService, Set<String>
      * @param originalCluster
      * @param endCluster
      */
-    public void diffRestartPulsarClient(Set<String> originalCluster, Set<String> endCluster) {
-        if (!originalCluster.equals(endCluster)) {
-            logger.info("pulsarConfig has changed, close current pulsarClientService and restart");
-            pulsarClientService.close();
-
-            pulsarCluster = configManager.getPulsarUrl2Token();
-            configManager.getPulsarConfig().setUrl2token(pulsarCluster);
-            pulsarClientService.initCreateConnection(this);
-            try {
-                initTopicSet(pulsarClientService, new HashSet<String>(topicProperties.values()));
-            } catch (Exception e) {
-                logger.info("pulsar sink restart, publish topic fail.", e);
-            }
+    public void diffUpdatePulsarClient(PulsarClientService pulsarClientService, Map<String, String> originalCluster,
+                                       Map<String, String> endCluster) {
+        MapDifference<String, String> mapDifference = Maps.difference(originalCluster, endCluster);
+        if (mapDifference.areEqual()) {
+            return;
+        }
 
+        logger.info("pulsarConfig has changed, close unused url clients and start new url clients");
+        Map<String, String> needToStart = new HashMap<>();
+        Map<String, String> needToClose = new HashMap<>();
+        needToClose.putAll(mapDifference.entriesOnlyOnLeft());
+        needToStart.putAll(mapDifference.entriesOnlyOnRight());
+        Map<String, MapDifference.ValueDifference<String>> differentToken = mapDifference.entriesDiffering();
+        for (String url : differentToken.keySet()) {
+            needToClose.put(url, originalCluster.get(url));
+            needToStart.put(url, endCluster.get(url));//token changed
         }
+
+        pulsarClientService.updatePulsarClients(this, needToClose, needToStart,
```

Comment:
It is not enough that pulsar client only use two status "start and close".
Closing status is necessary for waiting the completion of sending messages.

Reply:
There is a resend mechanism in sendMessage method. If its send operation is not completed when a pulsar client is closed, current event will be added into resendQueue to resend.

Reply:
The method "resendEvent" only is invoked in catchException and handleMessageSendException.
No catchException and handleMessageSendException when a pulsar client is closed.